test/test_linalg.py

```diff
@@ -2086,7 +2089,7 @@ def test_linalg_svd_compute_uv(self, device, dtype):
             # check linalg.svd vs numpy
             expected = np.linalg.svd(np_t, full_matrices, compute_uv=True)
             actual = torch.linalg.svd(t, full_matrices, compute_uv=True)
-            self.assertEqual(actual, expected)
+            # self.assertEqual(actual, expected)
```

Comment:
This test is wrong. SVD results for U and V are non-unique in many cases. Different backends may give different U and V results. See https://math.stackexchange.com/questions/644327/how-unique-are-u-and-v-in-the-singular-value-decomposition

We should focus on the svd properties, e.g. U and V are unitary, U*S*V^H is the same as the original matrix, sorted S diagonal elements are unique, etc. Those are already done in `test_svd` test.

Reply:
SGTM. @antocuni, @IvanYashchuk - does that sound right to you, too?

Reply:
It doesn't sound right to remove this test.
Satisfying the mathematical properties is of course the first priority, but also is NumPy (LAPACK) compatibility.
Yes, right and left singular vectors are not unique, similarly to as eigenvectors are not unique. However, they are usually normalized in the actual implementations, and then the difference is only in the sign of the vector.
Having the following checks:
```
self.assertEqual(abs(actual[0]), abs(expected[0]))
self.assertEqual(actual[1], expected[1])
self.assertEqual(abs(actual[2][:t.shape[0], :]), abs(expected[2][:t.shape[0], :]))
```
pass the tests. Alternatively, the sign of vectors can be made consistent as done [here](https://github.com/pytorch/pytorch/blob/master/test/test_linalg.py#L525-L536).

`[:t.shape[0], :]` indexing is required because cuSOLVER only computes the first `min(m, n)` right and left singular vectors leaving what is left as zeros. So in this test `actual[2][10:, :]` are zeros.
While LAPACK, MAGMA, and cuSOLVER's `gesvd` can compute all the singular vectors (when `jobu = jobvt = 'A'`, or `full_matrices=True` in Python).

This test shouldn't be removed, it needs slight modifications together with relevant notes in the documentation about the sign ambiguity (see [`eigh`'s doc](https://pytorch.org/docs/master/linalg.html#torch.linalg.eigh)) and `gesvdj` not computing all the matrix entries if `full_matrices=True`, I think even an error should be thrown in this case.

Reply:
Right, for real inputs, the output could be differed by a sign (\pm 1). However, for complex inputs, the output could be differed by an arbitrary phase (which is happening for cusolver gesvdj vs numpy lapack), e.g. `e^{i \phi}`, and this phase is different for each vector. Is there a standard of the phase that every implementation should use, or how should we unify the behavior for all phases for the numpy compatibility?

Reply:
There is no standard and it is not needed to satisfy signs or phases, however, this discrepancy should be documented.
The above checks that I wrote make tests pass both for real and complex input. In addition, I suggest adding checks similar to https://github.com/pytorch/pytorch/blob/master/test/test_linalg.py#L525-L536 where PyTorch and NumPy output are first transformed to follow the same sign convention and then values are compared.

Reply:
I think the issue here is the expectation of what the test is doing/should do, and it's the difference between white-box/black-box testing.
Both numpy's and pytorch's docs say that `svd` is computed by calling lapack's `_gesdd` function, so from some point of view the goal of this test is to ensure that we call `_gesdd` "in the same way as numpy", and thus we should expect bit-to-bit equal results, at least on CPU.
If a certain cuda backend returns different-but-correct results, I think it's fine to either adapt the test to allow a more relaxed check or to skip it completely.
Checking that the SVD properties hold should be done by other tests, IMHO.

Reply:
Looks like this test still needs to be updated per @IvanYashchuk's suggestion?

Reply:
I added doc change here https://github.com/pytorch/pytorch/pull/48436/commits/eedb71087f3bd0c4377770b69423f2464e388007. I think this line of test needs to be skipped, if not, I'm not sure how to modify test to satisfy the requirements. @IvanYashchuk  can you help?

Reply:
Ok, I agree now that this check could be removed since it says that the behavior is compared against NumPy whether we get "the very same behavior" but we can't get exactly the same output in NumPy and PyTorch.
Or you can use the following suggestion
```suggestion
            # sign/phase of the singular vectors is not unique and therefore absolute values are compared
            self.assertEqual(abs(actual[0]), abs(expected[0]))
            self.assertEqual(actual[1], expected[1])
            self.assertEqual(abs(actual[2][:t.shape[0], :]), abs(expected[2][:t.shape[0], :]))
```

Reply:
Modified as you suggested.